### ISSUE TITLE
`ParamSpec` within decorator is overly lenient

### ISSUE THREAD

**Issue by kwshi** [microsoft/pyright#4275]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Consider the following example.

```python
import typing

P = typing.ParamSpec("P")

def do_nothing(f: typing.Callable[[int], int]):
    def wrapped(*args: P.args, **kwargs: P.kwargs):
        return f(*args, **kwargs)

    return wrapped

@do_nothing
def my_func(n: int) -> int:
    return n + 1

my_func("hello") # should not type-check, but does
```

`do_nothing` doesn't change the functionality of `my_func` at all, since the `wrapped` function it returns simply invokes `f` as is. However the function signature of `my_func` is originally monomorphic (`(n: int) -> int`), while the `wrapped` function's signature is completely polymorphic (`P -> int`), which is overly generic, and so allows invocations like `my_func("hello", False)` and other stuff.

**To Reproduce**
Try example above.

**Expected behavior**
Should not type check. Error should be raised in the definition of the decorator at `f(*args, **kwargs)` (which tries to pass arbitrary paramspec as input to `f`, which can only take `[int]`).

**VS Code extension or command-line**
Neovim LSP extension + pyright 1.1.280.

**Comment by erictraut** [comment]
The current behavior is as designed.

The return type of `do_nothing` is not annotated, so pyright infers its type based on the types of the expressions returned by the function. In your example, there is only one return statement, and the evaluated type of `wrapped` is a function with unsolved type variables (in this case `P`). Unsolved type variables are effectively `Any`, so this matches any signature.

It's similar to this, which also doesn't generate an error.
```python
T = TypeVar("T")
def do_nothing(f: Callable[[int], int]):
    def wrapped(*args: T): pass
    return wrapped
```
